Message interviews:23 -  Read
From:    John Vlissides <S=vlis;OU=stanford;OU=interviews;ORG=edu;PRMD=switch;A
         DMD=arcom;C=ch>
To:      InterViews Mailing List <S=interviews;OU=stanford;OU=interviews;ORG=ed
         u;PRMD=switch;ADMD=arcom;C=ch>
Cc:      John Vlissides <S=vlis;OU=stanford;OU=lurch;ORG=edu;PRMD=switch;ADMD=a
         rcom;C=ch>
Subject: scroller patch

This patch makes scrollers behave better when the height/width of the
perspective they view is zero.

This patch is available for ftp as fix18.
---
Apply with ``cd iv; patch -p -N < diffs'' to patch files:

        src/libInterViews/scroller.c

*** /tmp/da8568	Tue Aug  1 08:18:55 1989
--- src/libInterViews/scroller.c	Tue Aug  1 08:17:32 1989
***************
*** 174,196 ****
  }

  void HScroller::GetBarInfo (register Perspective* s, Coord& left, int& width)
 {
      if (s->width == 0) {
! 	scale = 0.0;
      } else {
! 	scale = double(xmax + 1) / double(s->width);
      }
-     left = round(double(s->curx - s->x0) * scale);
-     width = max(round(double(s->curwidth) * scale), 2);
  }

! void VScroller::GetBarInfo (register Perspective* s, Coord& top, int& height)
 {
      if (s->height == 0) {
! 	scale = 0.0;
      } else {
! 	scale = double(ymax + 1) / double(s->height);
      }
-     top = round(double(s->cury - s->y0) * scale);
-     height = max(round(double(s->curheight) * scale), 2);
  }

  inline void HScroller::Bar (Coord x, int width) {
--- 174,204 ----
  }

  void HScroller::GetBarInfo (register Perspective* s, Coord& left, int& width)
 {
+     Coord maxwidth = xmax + 1;
+
      if (s->width == 0) {
!         scale = 0.0;
!         left = -1;
!         width = maxwidth + 2;
      } else {
! 	scale = double(maxwidth) / double(s->width);
!         left = round(double(s->curx - s->x0) * scale);
!         width = max(round(double(s->curwidth) * scale), 2);
      }
  }

! void VScroller::GetBarInfo (register Perspective* s, Coord& bot, int& height)
 {
!     Coord maxheight = ymax + 1;
!
      if (s->height == 0) {
!         scale = 0.0;
!         bot = -1;
!         height = maxheight + 2;
      } else {
! 	scale = double(maxheight) / double(s->height);
!         bot = round(double(s->cury - s->y0) * scale);
!         height = max(round(double(s->curheight) * scale), 2);
      }
  }

  inline void HScroller::Bar (Coord x, int width) {
***************
*** 238,250 ****
  }

  void VScroller::Redraw (Coord x1, Coord y1, Coord x2, Coord y2) {
!     Coord top;
      int height;

      Background(x1, y1, x2, y2);
!     GetBarInfo(shown, top, height);
!     Bar(top, height);
!     Outline(top, height);
  }

  void HScroller::Handle (Event& e) {
--- 246,258 ----
  }

  void VScroller::Redraw (Coord x1, Coord y1, Coord x2, Coord y2) {
!     Coord bot;
      int height;

      Background(x1, y1, x2, y2);
!     GetBarInfo(shown, bot, height);
!     Bar(bot, height);
!     Outline(bot, height);
  }

  void HScroller::Handle (Event& e) {
